src/os/bluestore/KernelDevice.cc

```diff
@@ -592,9 +592,15 @@ int KernelDevice::write(
 	   << dendl;
   assert(is_valid_io(off, len));
 
-  if ((!buffered || bl.get_num_buffers() >= IOV_MAX) &&
-      bl.rebuild_aligned_size_and_memory(block_size, block_size)) {
-    dout(20) << __func__ << " rebuilding buffer to be aligned" << dendl;
+  if ((!buffered || bl.get_num_buffers() >= IOV_MAX)) {
+    uint64_t aligned_size = block_size;
+    // Avoid after rebuild_aligned_size_and_memory, get_num_buffers() > IOV_MAX
+    if (bl.get_num_buffers() >= IOV_MAX && (len > (IOV_MAX * block_size))) {
+      aligned_size = ROUND_UP_TO(len / IOV_MAX, block_size);
+    }
+    if (bl.rebuild_aligned_size_and_memory(aligned_size, block_size)) {
+      dout(20) << __func__ << " rebuilding buffer to be aligned" << dendl;
```

Comment:
could you print out the `aligned_size` here also? i think it could help with the performance related diagnosis.